src/OpenTelemetry.AutoInstrumentation.StartupHook/StartupHook.cs

```diff
@@ -164,4 +167,51 @@ private static List<string> GetExcludedApplicationNames()
             return null;
         }
     }
+
+    private static void ThrowIfReferenceIncorrectOpenTelemetryVersion(string loaderAssemblyLocation)
+    {
+        var appReferenceAssemblies = Assembly.GetEntryAssembly()?.GetReferencedAssemblies();
+        if (appReferenceAssemblies == null)
+        {
+            return;
+        }
+
+        var index = Array.FindIndex(appReferenceAssemblies, assembly => assembly.Name == "OpenTelemetry");
+        bool throwException = false;
+
+        if (index != -1)
+        {
+            try
+            {
+                // Look up for type with an assembly name, will load the library.
+                // OpenTelemetry assembly load happens only if app has reference to the package.
+                var openTelemetryType = Type.GetType("OpenTelemetry.Sdk, OpenTelemetry");
+                if (openTelemetryType != null)
+                {
+                    var loadedOTelAssembly = Assembly.GetAssembly(openTelemetryType);
+                    var loadedOTelFileVersionInfo = FileVersionInfo.GetVersionInfo(loadedOTelAssembly?.Location);
+                    var loadedOTelfileVersion = new Version(loadedOTelFileVersionInfo.FileVersion);
+
+                    var profilerOTelLocation = Path.Combine(loaderAssemblyLocation, "OpenTelemetry.dll");
+                    var profilerOTelFileVersionInfo = FileVersionInfo.GetVersionInfo(profilerOTelLocation);
+                    var profilerOTelFileVersion = new Version(profilerOTelFileVersionInfo.FileVersion);
+
+                    if (loadedOTelfileVersion < profilerOTelFileVersion)
+                    {
+                        throwException = true;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                // Exception in evaluation should not throw or crash the process.
+                StartupHookEventSource.Log.Trace($"Couldn't evaluate reference to OpenTelemetry Sdk in an app. Exception: {ex}");
+            }
+        }
+
+        if (throwException)
+        {
+            throw new NotSupportedException("Application has direct or indirect reference to older version of OpenTelemetry package.");
```

Comment:
Is it possible to put loaded and expected versions here?

Reply:
Fixed 4b32407cc80fc3d4386c35d397d9497bb3c6e64b